File: MemberApp/Services/MemberService.cs

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using MemberApp.Models;
using Newtonsoft.Json;

namespace MemberApp.Services
{
    public class MemberService
    {
        private const string STORAGE_FILE = "members.json";
        
        public List<Member> GetAllMembers()
        {
            if (!File.Exists(STORAGE_FILE))
                return new List<Member>();
                
            string json = File.ReadAllText(STORAGE_FILE);
            return string.IsNullOrEmpty(json) 
                ? new List<Member>() 
                : JsonConvert.DeserializeObject<List<Member>>(json);
        }
        
        public Member AddMember(Member member)
        {
            var members = GetAllMembers();
            member.Id = DateTime.Now.Ticks.ToString();
            members.Add(member);
            SaveMembers(members);
            return member;
        }
        
        public void DeleteMember(string id)
        {
            var members = GetAllMembers();
            var updatedMembers = members.Where(m => m.Id != id).ToList();
            SaveMembers(updatedMembers);
        }
        
        public List<Member> SearchMembers(string query)
        {
            if (string.IsNullOrWhiteSpace(query))
                return GetAllMembers();
                
            var members = GetAllMembers();
            query = query.ToLower();
            
            return members.Where(m => 
                m.Name.ToLower().Contains(query) ||
                m.Email.ToLower().Contains(query) ||
                m.Phone.Contains(query) ||
                m.Department.ToLower().Contains(query) ||
                m.Rank.ToLower().Contains(query)
            ).ToList();
        }
        
        public string ExportMembersToCSV()
        {
            var members = GetAllMembers();
            if (members.Count == 0)
                return string.Empty;
                
            var headers = new[] { "الاسم", "رقم الهاتف", "البريد الإلكتروني", "تاريخ الانضمام", "الرتبة", "القسم" };
            var sb = new StringBuilder();
            
            sb.AppendLine(string.Join(",", headers));
            
            foreach (var member in members)
            {
                sb.AppendLine($"{member.Name},{member.Phone},{member.Email},{member.JoinDate},{member.Rank},{member.Department}");
            }
            
            return sb.ToString();
        }
        
        private void SaveMembers(List<Member> members)
        {
            string json = JsonConvert.SerializeObject(members, Formatting.Indented);
            File.WriteAllText(STORAGE_FILE, json);
        }
    }
}

File: src/services/MemberService.ts

import { Member } from "@/models/Member";

// محاكاة لقاعدة بيانات محلية باستخدام localStorage
class MemberService {
  private readonly STORAGE_KEY = 'members';

  // جلب جميع الأعضاء
  getAllMembers(): Member[] {
    const members = localStorage.getItem(this.STORAGE_KEY);
    return members ? JSON.parse(members) : [];
  }

  // إضافة عضو جديد
  addMember(member: Omit<Member, 'id'>): Member {
    const members = this.getAllMembers();
    const newMember = {
      ...member,
      id: Date.now().toString()
    };
    
    localStorage.setItem(this.STORAGE_KEY, JSON.stringify([...members, newMember]));
    return newMember;
  }

  // حذف عضو
  deleteMember(id: string): void {
    const members = this.getAllMembers();
    const updatedMembers = members.filter(member => member.id !== id);
    localStorage.setItem(this.STORAGE_KEY, JSON.stringify(updatedMembers));
  }

  // البحث عن أعضاء
  searchMembers(query: string): Member[] {
    if (!query.trim()) return this.getAllMembers();
    
    const members = this.getAllMembers();
    const lowercaseQuery = query.toLowerCase();
    
    return members.filter(member => 
      member.name.toLowerCase().includes(lowercaseQuery) ||
      member.email.toLowerCase().includes(lowercaseQuery) ||
      member.phone.includes(query) ||
      member.department.toLowerCase().includes(lowercaseQuery) ||
      member.rank.toLowerCase().includes(lowercaseQuery)
    );
  }

  // تصدير الأعضاء إلى ملف CSV
  exportMembersToCSV(): string {
    const members = this.getAllMembers();
    if (members.length === 0) return '';
    
    const headers = ['الاسم', 'رقم الهاتف', 'البريد الإلكتروني', 'تاريخ الانضمام', 'الرتبة', 'القسم'];
    const csvRows = [
      headers.join(','),
      ...members.map(member => [
        member.name,
        member.phone,
        member.email,
        member.joinDate,
        member.rank,
        member.department
      ].join(','))
    ];
    
    return csvRows.join('\n');
  }
}

export const memberService = new MemberService();
